### ISSUE TITLE
Use something other than Antlr

### ISSUE THREAD

**Issue by joemcbride** [graphql-dotnet/graphql-dotnet#2]
It has been suggested that the dependency on Antlr is off-putting due to the possible Java/IKVM dependency (note that the NuGet package doesn't have a Java/IKVM dependency, just the development project has that dependency).  I don't have plans on switching currently, though if there is enough interest in doing so that is certainly an option.

Possible replacements:

* https://github.com/sprache/Sprache

**Comment by PeteDuncanson** [comment]
Hi Joe,

I was tempted to go with ANTLR with our version of this but I just don't think it was a good fit. The JS reference version of GraphQL will continually get updated with any fixes so I went the route of porting their parser. My thinking was it would be easier for people to read that and translate it into C# fixes in a parser so we've have more eyes on it.

You've done such a great amount of work on this I'm happy to bring my code to the table and add it to this project and put our heads together as you said on twitter ;)

I'll create a fork and see about getting my stuff added. Not got too much but the parser is about 40% done I think. If I focus on that bit we could swap it out for Antlr once its up and running assuming you are happy with it.

Cheers

Pete

**Comment by joemcbride** [comment]
Hey Pete, thanks for offering to help!

My thinking with going with ANTLR was so that I could get something working fairly quickly, and overall I think they do a good job at solving the lexer/parser problem.

I think going with a custom solution would be fine, though from experience that can be hard to get right.  As you said though, that could possibly elicit more help if it is all in .NET.

I'll add in some hooks to make it easier to swap out the `Document` builder.  That way you can run the existing tests to see if its working.

**Comment by joemcbride** [comment]
Added some interfaces to make it easier to replace existing classes.

Implement `IDocumentBuilder` which takes in the query data and returns a `Document`.

```csharp
public interface IDocumentBuilder
{
    Document Build(string data);
}

public class MyCustomBuilder : IDocumentBuilder
{
    ...
}
```

You can then replace the builder in the Tests by making this change:

```csharp
public class StarWarsBasicQueryTests
    : QueryTestBase<StarWarsSchema, MyCustomBuilder>
{
    ...
}
```

**Comment by joemcbride** [comment]
How about Facebook's C++ parser?  https://github.com/graphql/libgraphqlparser

**Comment by bryanerayner** [comment]
It seems like using the C++ implementation would be mean a lot less work in the future.

**Comment by attilah** [comment]
For parser generator I'd take a look at Jetbrains' Nitra.

**Comment by zhech2** [comment]
I ported facebook's javascript version to c#.  Any interest in including it?

**Comment by attilah** [comment]
Wow ;-) can you share a link? :dancer:

**Comment by joemcbride** [comment]
Worth a look for sure.

**Comment by zhech2** [comment]
Here it is: https://github.com/zhech2/graphql-dotnet/commit/4f0c720569fe76f9c7175f269775819b113d9ebf.

It could still use work but at least you can review it.  So far, I've only written enough to get all tests passing.  Some of the code to convert from the parser's ASTs to the current format is not implemented.  This is because the current tests were not calling any of the methods and I didn't want to implement them until I write tests.

If this method were used, I was thinking that the original JS source could be saved or referenced in some manner, so that it could be compared against.  This would allow for updating the C# when the JS source is updated.

**Comment by baronfel** [comment]
@zhech2 you could potentially use [Paket's Github dependencies feature](https://fsprojects.github.io/Paket/github-dependencies.html) to reference the specific parser.js file in that repo at a point in time.  Then, any updates to that would show as an updateable dependency when 'paket update' is run.

**Comment by joemcbride** [comment]
Closed with the merging of #127
